### ISSUE TITLE
Wrong type inferred in `finally` block

### ISSUE THREAD

**Issue by not-my-profile** [microsoft/pyright#3105]
For the following code pyright should report an error but it doesn't. In #2930 you said that pyright gets this wrong but in my testing it doesn't.

```python
def foo() -> None:
    file = None
    try:
        raise ValueError()
        file = open('test.json')
    except Exception:
        return None
    finally:
        file.name # error is None but pyright doesn't complain
```

**Comment by not-my-profile** [comment]
I was referring to your sentence:

>Incidentally, mypy doesn't get this right either. It claims that the type of `file` within the finally clause is never `None`, but clearly it can be `None`

In my testing both mypy and pytype get it right, whereas pyright gets it wrong.

**Comment by erictraut** [comment]
This was a regression from the previous change introduced for #2930. I'm backing out the previous change. 

The situation described in #2930 is an edge case that we won't be able to handle for now in pyright. You'll need to find a workaround such as an assert statement.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.225, which I just published. It will also be included in the next release of pylance.
